tests/StubsTypeHintsTest.php

```diff
@@ -16,225 +16,170 @@
     }
 
-    private static function checkMethodDoesNotHaveReturnTypeHints(int $sinceVersion, PHPClass $parentClass, PHPFunction $function)
+    /**
+     * @dataProvider \StubTests\TestData\Providers\StubsTestDataProviders::stubMethodWithoutUnionReturnTypeHintsProvider
+     * @param PHPMethod $stubMethod
+     */
+    public function testMethodDoesNotHaveUnionReturnTypeHint(PHPMethod $stubMethod)
     {
-        $returnTypeHint = $function->returnType;
-        if ($sinceVersion < 7) {
-            if (!$function->hasMutedProblem(StubProblemType::FUNCTION_HAS_RETURN_TYPEHINT)) {
-                self::assertEmpty($returnTypeHint, "Method '{$parentClass->name}::{$function->name}' has since version '$sinceVersion'
-            but has return typehint '$returnTypeHint' that supported only since PHP 7. Please declare return type via PhpDoc");
-            } else {
-                self::markTestSkipped('Method ignored');
-            }
-        } else {
-            self::markTestSkipped("Function '{$function->name}' has since version > 7");
-        }
+        $sinceVersion = StubsTypeHintsTest::getFirstSinceVersion($stubMethod);
+        self::assertFalse(str_contains($stubMethod->returnType, '|'),
+            "Method '{$stubMethod->parentName}::{$stubMethod->name}' has since version '$sinceVersion'
+            but has union return typehint '$stubMethod->returnType' that supported only since PHP 8.0. 
+            Please declare return type via PhpDoc");
     }
 
-    private static function checkMethodDoesNotHaveScalarTypeHints(int $sinceVersion, PHPClass $parentClass, PHPFunction $function)
+    private static function compareTypeHintsWithReflection(PHPParameter $parameter, PHPParameter $stubParameter, ?string $functionName): void
     {
-        if ($sinceVersion < 7) {
-            if (empty($function->parameters)) {
-                self::markTestSkipped('Parameters list empty');
-            } else {
-                foreach ($function->parameters as $parameter) {
-                    if (!$parameter->hasMutedProblem(StubProblemType::PARAMETER_HAS_SCALAR_TYPEHINT)) {
-                        self::assertFalse($parameter->type === 'int' || $parameter->type === 'float' ||
-                            $parameter->type === 'string' || $parameter->type === 'bool',
-                            "Method '{$parentClass->name}::{$function->name}' with @since '$sinceVersion'  
-                has parameter '{$parameter->name}' with typehint '{$parameter->type}' but typehints available only since php 7");
-                    } else {
-                        self::markTestSkipped('Parameter ignored');
-                    }
-                }
-            }
-        } else {
-            self::markTestSkipped("Function '{$function->name}' has since version > 7");
+        $unifiedReflectionParameter = preg_replace('/\?/', 'null|', $parameter->type);
+        $unifiedStubsParameter = preg_replace('/\?/', 'null|', $stubParameter->type);
```

Comment:
Do we really want this? It will allow: `string|?int` in stubs.

Reply:
Hm...good point. The idea was to pass cases when reflections returns `?int` but stub declared as `null|int`. So it requires the check that there is just one parameter type. I'll update this